[Metadata Immutability] Change different indices lookup objects from array type to lists

akolarkunnu opened the pull request:
### Description
Changed the arrays to immutable List instances, added new versions of the getters which returns List instances.

### Related Issues
Resolves #8647

Signed-off-by: Abdul Muneer Kolarkunnu [muneer.kolarkunnu@netapp.com]

### Check List
- [x] New functionality includes testing.
- [x] All tests pass
- [ ] ~New functionality has been documented.~
- [x] New functionality has javadoc added
- [ ] ~API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).~
- [ ] ~Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))~
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))~
- [ ] ~Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

vikasvb90 commented on server/src/main/java/org/opensearch/cluster/metadata/Metadata.java:
```diff
@@ -1576,12 +1576,12 @@ protected Metadata buildMetadataWithPreviousIndicesLookups() {
-                Arrays.copyOf(previousMetadata.visibleOpenIndices, previousMetadata.visibleOpenIndices.length),
-                Arrays.copyOf(previousMetadata.allClosedIndices, previousMetadata.allClosedIndices.length),
-                Arrays.copyOf(previousMetadata.visibleClosedIndices, previousMetadata.visibleClosedIndices.length),
+                previousMetadata.allIndices,
```
Why is this also not a copy? Places where we create temporary metadata objects and mutate may lead to issues.

akolarkunnu replied:
Previously these data structures were arrays. Now these are immutable collections(unmodifiable collection). So these copying are not required.

sandeshkr419 commented:
Thanks @akolarkunnu, changes look good to me!
@andrross @shwetathareja Do you have any additional comments on this?

andrross commented:
@akolarkunnu Looks good. Unfortunately there's a merge conflict now. Can you rebase?

shwetathareja commented on server/src/main/java/org/opensearch/action/admin/cluster/health/ClusterHealthResponse.java:
```diff
@@ -215,13 +216,13 @@ public ClusterHealthResponse(StreamInput in) throws IOException {
     public ClusterHealthResponse(
         String clusterName,
-        String[] concreteIndices,
+        Set<String> concreteIndices,
```
changes in ClusterHealthResponse will result in backward incompatibility

akolarkunnu replied:
Yes, this is a breaking change, it's already discussed in https://github.com/opensearch-project/OpenSearch/pull/14557 . Our plan is to have breaking changes in 3.x and non-breaking changes separately for 2.x.

andrross replied:
This approach is fine, but the way I've usually done it is to first do the backwards compatible change on main and backport it, then follow up with another PR to remove the backward compatible code on main.

shwetathareja replied:
I missed the point but what is the need for changing it to Set in Response?

akolarkunnu replied:
Our aim is to reduce the conversion String[] to List and vice versa. If we keep String[] in here we have to convert String[] to List before creating ClusterStateHealth. If we don't make change in ClusterStateHealth, then inside the constructor, we have to convert List(return value of clusterState.metadata().getConcreteAllIndices()) to String[]. 
To avoid this kind of to and fro conversion, we did this change.

shwetathareja commented on server/src/main/java/org/opensearch/cluster/metadata/IndexNameExpressionResolver.java:
```diff
@@ -1182,17 +1182,17 @@ private boolean isEmptyOrTrivialWildcard(List<String> expressions) {
         private static List<String> resolveEmptyOrTrivialWildcard(IndicesOptions options, Metadata metadata) {
             if (options.expandWildcardsOpen() && options.expandWildcardsClosed() && options.expandWildcardsHidden()) {
-                return Arrays.asList(metadata.getConcreteAllIndices());
+                return List.copyOf(metadata.getConcreteAllIndices());
```
This would be expensive operation to copy the set to a list every time for index resolution.
Have run micro benchmark on your changes?

akolarkunnu replied:
Here ConcreteAllIndices(allIndices) are unmodifiableSet, so List.copyOf() will not really create a copy . 
There is an implementation Note in javadoc of this API - "If the given Collection is an unmodifiable List, calling copyOf will generally not create a copy". Also if we look in to the internal implementation of Lisy.copyOf() it's clear that, it's not creating a copy if collection is an AbstractImmutableList.

shwetathareja replied:
you are doing
```Collections.unmodifiableSet(allIndices);``` which creates new object of ```UnmodifiableSet```. ```UnmodifiableSet``` extends ```UnmodifiableCollection``` and it is not extending AbstractImmutableList. It will end copying the elements. Can you cross check?

akolarkunnu replied:
Yes, you are correct. So what do you suggest here?
concrete indices treated as Set in the existing implementation - https://github.com/opensearch-project/OpenSearch/blob/67a2e4c7275afa93ce2c6fc2107ca0f7a8c461bd/server/src/main/java/org/opensearch/cluster/metadata/Metadata.java#L1636
That's why I kept only concrete indices as Set and all other indices as List. If we keep concrete indices also as List, then this conversion won't be required.

shwetathareja replied:
yes right, lets check the usage in the code if it has to be Set, else i agree keeping it as list would be better.

akolarkunnu replied:
From the usage wise, it kept as Set to avoid duplicates in "indices" and to have better performance during "retainAll()". 
From the java doc of retainAll() - "If the specified collection is also a set, this operation effectively modifies this set so that its value is the intersection of the two sets."

So I don't think we can use List instead of Set here.

andrross replied:
How do we know there is not a regression here, given that `Arrays.asList` was previously a lightweight operation but this is now a per-element copy?

akolarkunnu replied:
@sandeshkr419 @andrross @shwetathareja Please review

rajiv-kv replied:
> I performed the performance testing as @sandeshkr419 suggested below. I created an index and 50000 aliases initially, then again created 500*10 = 5000 aliases (so total 55000) and calculated the average time taken for these last 500 sets of alias creation separately. The results shows that there is a consistent improvement in each sets(~10ms) after this change.


Thank @akolarkunnu for performing the perf runs. Since there is a change in ClusterStateHealth, we might need a similar run for `health` API calls. Could we also capture the memory allocations corresponding to the duration of run and compare them before / after change ? 

Also, trying to understand as to why 50K aliases were created and not the actual index ?

sandeshkr419 replied:
> Also, trying to understand as to why 50K aliases were created and not the actual index ?

We could also create indices, its just you need to modify node limits on max indices/shards per node - plus the the change is cluster state is identical (as in same/similar objects are modified). 


> Since there is a change in ClusterStateHealth, we might need a similar run for health API calls. Could we also capture the memory allocations corresponding to the duration of run and compare them before / after change ?

I don't think we would be able to capture changes (improvements actually) here, the only thing we have omitted here is back and worth list-set conversions for some objects (stored as one type, retrieved as different). [Recalling from my own PR when I had tried to benchmark some changes in cluster state]

rajiv-kv replied:
Health API iterates over the IndexMetadata to compute the overall cluster health. Hence i was thinking that creating indexes rather than alias would have more entries to iterate for health computation. For cluster health computation, there were less memory allocations seen when it was switched to size based array iteration. Ref PR :- https://github.com/opensearch-project/OpenSearch/pull/15492

Please take a look at the memory allocations in flame graph in the referenced PR. Could we verify that switching from array to set iterator does not incur additional memory allocations for health API.

sandeshkr419 replied:
@akolarkunnu Can you please take these changes to closure by profiling health API as in https://github.com/opensearch-project/OpenSearch/pull/15492 

That should bring confidence on whether or not to get these changes in code or not.

akolarkunnu replied:
Ok, I will resume this work soon.

akolarkunnu replied:
I ran performance test by crating indices using a script, which ran for 2 hours and 3 minutes. On comparing with changes and without changes, with changes I am consistently able to create ~30 more indices. So it is clearly a performance improvement.
Also I captured FlameGraph of 5 mins after 2 hours of index creation by using async profiler, and I don't see anything unusual there. Attaching both cpu and alloc with and without changes.
[flamegraphs.zip](https://github.com/user-attachments/files/18556047/flamegraphs.zip)

shwetathareja replied:
@akolarkunnu : Thanks for benchmarking the creation times. Can you please benchmark ```resolveEmptyOrTrivialWildcard``` method with and without your change. This resolution would happen in the context of every indexing/ search context which could turn out to be very expensive with this copy operation

sandeshkr419 replied:
@shwetathareja A quick clarification - so if `resolveEmptyOrTrivialWildcard` is called in every search/indexing request - will the indexing/search benchmarks suffice to establish the impact? 

If yes, please checkout the comparison results - https://github.com/opensearch-project/OpenSearch/pull/14723#issuecomment-2341971638 - I do see a good improvement in reduction of search and indexing latencies, increase in majority of indexing/search throughputs.

One catch I see is that the indices+aliases size is relatively low in the OSB benchmark, but I guess the slight increase in performance we see can be accounted for the reduction in conversions back and forth to set/lists. What do you think?

shwetathareja replied:
OSB benchmark may not show impact as it would run with handful indices.

akolarkunnu commented on server/src/main/java/org/opensearch/index/recovery/RemoteStoreRestoreService.java:
```diff
@@ -165,7 +165,7 @@ public RemoteRestoreResult restore(
         } else {
             List<String> filteredIndices = filterIndices(
-                List.of(currentState.metadata().getConcreteAllIndices()),
+                List.copyOf(currentState.metadata().getConcreteAllIndices()),
```
No, it can operate only on List

akolarkunnu commented:
{"run-benchmark-test": "id_9"}

rishabh6788 commented:
not sure why the benchmark approval didn't run. running again.

rishabh6788 commented:
{"run-benchmark-test": "id_9"}

sandeshkr419 reviewed (approved):
The performance testing as per https://github.com/opensearch-project/OpenSearch/pull/14723/files#r1752147549 gives confidence that we are not introducing any regression. Also, we avoid back and forth changes with lists/sets now. Changes look good.

andrross commented:
Thanks for sticking with us @akolarkunnu! The DCO check is failing, I think related to the fact that you've got 354 commits in this PR. I'd suggest squashing/rebasing everything to just your single commit on top of the latest commit on main. Also, you're currently signing your commit with `33829651+akolarkunnu@users.noreply.github.com`. Please take a look at our [DCO policy](https://github.com/opensearch-project/.github/blob/main/CONTRIBUTING.md#developer-certificate-of-origin). We'd appreciate it if you could sign the commit with a valid email address.

akolarkunnu committed:
[Metadata Immutability] Change different indices lookup objects from array type to lists

Changed the arrays to immutable List instances, added new versions of the getters which returns List instances.

Resolves #8647

Signed-off-by: Abdul Muneer Kolarkunnu <muneer.kolarkunnu@netapp.com>

akolarkunnu commented:
> Thanks for sticking with us @akolarkunnu! The DCO check is failing, I think related to the fact that you've got 354 commits in this PR. I'd suggest squashing/rebasing everything to just your single commit on top of the latest commit on main. Also, you're currently signing your commit with `33829651+akolarkunnu@users.noreply.github.com`. Please take a look at our [DCO policy](https://github.com/opensearch-project/.github/blob/main/CONTRIBUTING.md#developer-certificate-of-origin). We'd appreciate it if you could sign the commit with a valid email address.

Corrected.

andrross reviewed (approved):
Looks good to me! @shwetathareja Can you take another quick pass?
